### DISCUSSION TITLE
Add @app.before_request and @app.after_request decorators to FastAPI

### DISCUSSION THREAD

**Discussion by StrawIII** [fastapi/fastapi#11258] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.before_request
def before_request(request: Request):
    # do something with the request
    return request
```

### Description

The `@app.before_request` decorator would allow developers to register functions to be run before each request is processed. This could be useful for tasks such as request logging, authentication, or loading user-specific data. Similarly, the `@app.after_request` decorator would enable functions to run after a request has been processed but before sending the response to the client. This is ideal for modifying the response, adding headers, or performing cleanup activities. Current implementation using @app.middleware seems little too complicated and also it is clucky to acceess for example request and response body.

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.8.10

**Comment by JavierSanchezCastro** [comment]
You can do both operations using middleware:

```python3
from fastapi import FastAPI, Request
from starlette.responses import Response

app = FastAPI()

@app.middleware("http")
async def some_middleware(request: Request, call_next):
    #Call functions here
    #...
    response = await call_next(request)
    #Get Response
    response_body = b""
    async for chunk in response.body_iterator:
        response_body += chunk
    print(f"response_body={response_body.decode()}")
    return Response(content=response_body, status_code=response.status_code, 
        headers=dict(response.headers), media_type=response.media_type)

@app.get("/")
async def read_main():
    return {"message": "Hello World"}

```
Code from https://stackoverflow.com/questions/71882419/fastapi-how-to-get-the-response-body-in-middleware

**Comment by VatsalJagani** [comment]
@StrawIII - Middleware is the answer.
https://fastapi.tiangolo.com/tutorial/middleware/

**Comment by YuriiMotov** [comment, marked as answer]
If it can easily be done using existing features, it will unlikely be added to FastAPI.

As was mentioned, [Middlewares](https://fastapi.tiangolo.com/tutorial/middleware/) is one of conventional ways to add such functionality.

Another way is to use dependency with `yield`:
```python

from fastapi import Depends, FastAPI, Request

app = FastAPI()

async def wrap_request(request: Request):
    print("Do something before request")
    try:
        yield
    finally:
        print("Do something after request")


@app.get("/", dependencies=[Depends(wrap_request)])
async def read_root():
    return {"message": "Hello, World!"}

```
